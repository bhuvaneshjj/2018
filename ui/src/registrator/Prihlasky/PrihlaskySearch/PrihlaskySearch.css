.PrihlaskySearch {
  width: 25em;
}