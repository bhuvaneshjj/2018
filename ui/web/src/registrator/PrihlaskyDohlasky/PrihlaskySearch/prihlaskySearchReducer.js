import { AKTUALNI_ROK } from '../../../constants';
import { prijmeniJmenoNarozeniSortMethod } from '../../../sort';

export const getSearchOptions = ({ ucastnici }) => {
  const selected = ucastnici.allIds.map((id) => {
    const ucastnik = ucastnici.byIds[id];
    const posledniRok = ucastnik.roky[0];
    const { prijmeni, jmeno, narozeni } = ucastnik[posledniRok].udaje;
    const result = { id, prijmeni, jmeno, narozeni };
    if (ucastnik[AKTUALNI_ROK]) {
      result.kod = ucastnik[AKTUALNI_ROK].prihlaska.kod;
    }
    return result;
  });
  return selected.sort(prijmeniJmenoNarozeniSortMethod);
};
