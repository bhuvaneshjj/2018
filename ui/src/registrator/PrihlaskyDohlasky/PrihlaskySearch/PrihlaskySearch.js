import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Typeahead } from 'react-bootstrap-typeahead';
import './PrihlaskySearch.css';

const stripDiactrics = string => string.normalize('NFD').replace(/[\u0300-\u036f]/g, '');

const isMatch = (option, { text }) => {
  const { prijmeni, kod } = option;

  const prijmeniLower = stripDiactrics(prijmeni.toLowerCase());
  const textLower = stripDiactrics(text.toLowerCase());

  return prijmeniLower.startsWith(textLower) || kod === text;
};

class PrihlaskySearch extends PureComponent {
  componentDidMount = () => {
    if (process.env.NODE_ENV !== 'test') {
      this.typeahead.focus();
    }
  };

  render = () => {
    const { options, onSelect } = this.props;

    return (
      <Typeahead
        autoComplete={false}
        className="PrihlaskySearch"
        emptyLabel="Nic nenalezeno."
        filterBy={isMatch}
        highlightOnlyResult={true}
        labelKey="prijmeni"
        minLength={1}
        options={options}
        placeholder="Začni psát příjmení nebo vlož kód přihlášky."
        ref={ref => {
          if (ref) {
            this.typeahead = ref.getInstance();
          }
        }}
        selectHintOnEnter={true}
        onChange={results => {
          if (results.length >= 1) {
            // Typeahead provides an array so take the first result.
            onSelect(results[0]);
          }
        }}
        renderMenuItemChildren={option => {
          const { prijmeni, jmeno, narozeni } = option;
          return `${prijmeni} ${jmeno}, ${narozeni.rok}`;
        }}
      />
    );
  };
}

PrihlaskySearch.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      prijmeni: PropTypes.string.isRequired,
      jmeno: PropTypes.string.isRequired,
      narozeni: PropTypes.shape({
        rok: PropTypes.number.isRequired
      }).isRequired,
      kod: PropTypes.string
    })
  ).isRequired,
  onSelect: PropTypes.func.isRequired
};

export default PrihlaskySearch;
