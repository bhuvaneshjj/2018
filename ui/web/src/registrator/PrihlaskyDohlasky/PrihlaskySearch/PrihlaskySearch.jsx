import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Typeahead } from 'react-bootstrap-typeahead';
import './PrihlaskySearch.css';

const stripDiactrics = string => string.normalize('NFD').replace(/[\u0300-\u036f]/g, '');

const isMatch = (option, { text }) => {
  const { prijmeni, kod } = option;

  const prijmeniLower = stripDiactrics(prijmeni.toLowerCase());
  const textLower = stripDiactrics(text.toLowerCase());

  return prijmeniLower.startsWith(textLower) || kod === text;
};

const PrihlaskySearch = ({ options, onSelect }) => {
  const typeaheadRef = useRef();

  useEffect(() => {
    if (process.env.NODE_ENV !== 'test') {
      typeaheadRef.current.focus();
    }
  });

  return (
    <Typeahead
      autoComplete={false}
      className="PrihlaskySearch"
      emptyLabel="Nic nenalezeno."
      filterBy={isMatch}
      highlightOnlyResult={true}
      labelKey="prijmeni"
      minLength={1}
      options={options}
      placeholder="Začni psát příjmení nebo vlož kód přihlášky."
      ref={typeaheadRef}
      selectHintOnEnter={true}
      onChange={results => {
        if (results.length >= 1) {
          // Typeahead provides an array so take the first result.
          onSelect(results[0]);
        }
      }}
      renderMenuItemChildren={option => {
        const { prijmeni, jmeno, narozeni } = option;
        return `${prijmeni} ${jmeno}, ${narozeni.rok}`;
      }}
    />
  );
};

PrihlaskySearch.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      prijmeni: PropTypes.string.isRequired,
      jmeno: PropTypes.string.isRequired,
      narozeni: PropTypes.shape({
        rok: PropTypes.number.isRequired
      }).isRequired,
      kod: PropTypes.string
    })
  ).isRequired,
  onSelect: PropTypes.func.isRequired
};

export default PrihlaskySearch;
