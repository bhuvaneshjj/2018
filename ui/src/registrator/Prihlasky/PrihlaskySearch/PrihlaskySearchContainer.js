import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getSearchOptions } from './prihlaskySearchReducer';
import PrihlaskySearch from './PrihlaskySearch';
import { ucastnikSelected } from '../PrihlaskyActions';

const mapStateToProps = state => {
  const { entities } = state;

  return {
    entities,
    options: getSearchOptions(entities)
  };
};

const mapDispatchToProps = dispatch => ({ dispatch });

const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const { entities, options } = stateProps;
  const { dispatch } = dispatchProps;
  const { routeOnSelect } = ownProps;

  return {
    options,
    onSelect: async option => {
      await dispatch(ucastnikSelected({ ...option, ...entities }));
      routeOnSelect(option.id);
    }
  };
};

const PrihlaskySearchContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(
  PrihlaskySearch
);

PrihlaskySearchContainer.propTypes = {
  routeOnSelect: PropTypes.func.isRequired
};

export default PrihlaskySearchContainer;
