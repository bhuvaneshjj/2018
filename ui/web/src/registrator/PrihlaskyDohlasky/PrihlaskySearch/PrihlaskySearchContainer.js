import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getSearchOptions } from './prihlaskySearchReducer';
import PrihlaskySearch from './PrihlaskySearch';

const mapStateToProps = ({ entities }) => ({
  options: getSearchOptions(entities),
});

const mapDispatchToProps = (dispatch, ownProps) => {
  const { routeOnSelect } = ownProps;

  return {
    onSelect: (option) => routeOnSelect(option.id),
  };
};

const PrihlaskySearchContainer = connect(mapStateToProps, mapDispatchToProps)(PrihlaskySearch);

PrihlaskySearchContainer.propTypes = {
  routeOnSelect: PropTypes.func.isRequired,
};

export default PrihlaskySearchContainer;
