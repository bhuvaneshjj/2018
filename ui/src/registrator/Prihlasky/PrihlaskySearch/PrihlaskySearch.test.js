import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import PrihlaskySearch from './PrihlaskySearch';

const options = [
  { id: '===id1===', prijmeni: 'Hudec', jmeno: 'Jiří', narozeni: { rok: 1956 }, kod: '===kod1===' },
  { id: '===id2===', prijmeni: 'Suchá', jmeno: 'Jana', narozeni: { rok: 1967 }, kod: '===kod2===' },
  { id: '===id3===', prijmeni: 'Kryl', jmeno: 'Jan', narozeni: { rok: 2001 } }
];

it('renders', () => {
  const component = renderer.create(
    <PrihlaskySearch options={options} onChange={jest.fn()} onSelect={jest.fn()} />
  );

  // Typeahead component contains dynamic 'aria-owns' prop. This destroys snapshot testing.
  const json = component.toJSON();
  const input = json.children[0].children[0].children[0].children[0];
  expect(input.type).toEqual('input');
  expect(input.props['aria-owns']).toEqual(expect.stringContaining('rbt-menu-'));
  input.props['aria-owns'] = 'rbt-menu-*';
  expect(json).toMatchSnapshot();
});

it('handle select', () => {
  const onSelect = jest.fn();

  const wrapper = mount(
    <PrihlaskySearch options={options} onChange={jest.fn()} onSelect={onSelect} />
  );
  expect(wrapper.find('.rbt-input-main')).toHaveLength(1);

  wrapper.find('.rbt-input-main').simulate('change', { target: { value: 'hu' } });
  wrapper.find('.rbt-input-main').simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 });
  expect(onSelect).toHaveBeenCalledWith(options[0]);
});
