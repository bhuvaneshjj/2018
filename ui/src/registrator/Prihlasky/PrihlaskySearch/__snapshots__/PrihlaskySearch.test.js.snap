// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="rbt open clearfix PrihlaskySearch"
  style={
    Object {
      "position": "relative",
    }
  }
  tabIndex={-1}
>
  <div
    className="rbt-input form-control"
    disabled={false}
    onClick={[Function]}
    onFocus={[Function]}
    tabIndex={-1}
  >
    <div
      className="rbt-input-wrapper"
    >
      <div
        style={
          Object {
            "display": "block",
            "position": "relative",
          }
        }
      >
        <input
          aria-activedescendant=""
          aria-autocomplete="both"
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-owns="rbt-menu-ehtlnc8qfzm"
          autoComplete="off"
          className="rbt-input-main"
          disabled={false}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Začni psát příjmení nebo vlož kód přihlášky."
          role="combobox"
          style={
            Object {
              "backgroundColor": "transparent",
              "border": 0,
              "boxShadow": "none",
              "cursor": "inherit",
              "outline": "none",
              "padding": 0,
              "width": "100%",
            }
          }
          type="text"
          value=""
        />
        <div
          className={undefined}
          style={
            Object {
              "bottom": 0,
              "display": "block",
              "pointerEvents": "none",
              "position": "absolute",
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <input
            aria-hidden={true}
            className="rbt-input-hint"
            style={
              Object {
                "backgroundColor": "transparent",
                "border": 0,
                "boxShadow": "none",
                "boxSizing": "content-box",
                "color": "rgba(0, 0, 0, 0.35)",
                "cursor": "inherit",
                "outline": "none",
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex={-1}
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="sr-only rbt-sr-status"
    role="status"
  >
    0 selections
  </div>
</div>
`;
